doc/manual/src/glossary.md

```diff
@@ -115,14 +115,14 @@
     then be built.
 
   - [reference]{#gloss-reference}\
-    A store path `P` is said to have a reference to a store path `Q` if
-    the store object at `P` contains the path `Q` somewhere. The
-    *references* of a store path are the set of store paths to which it
-    has a reference.
+    A [store object] `O` is said to have a *reference* to a store object `P` if the [store path] of `P` appears in the contents of `O`.
+    The *references* of a store object `O` are the set of store objects to which `O` has a reference.
 
-    A derivation can reference other derivations and sources (but not
-    output paths), whereas an output path only references other output
-    paths.
+    Source files have no references.
+    A [store derivation] can only reference source files and other store derivations, including itself.
+    In contrast, a store object that was produced from a [derivation] can only reference other "derived" store objects.
```

Comment:
I wouldn't get too much into the source file vs build result distinction here. It is not a very deep thing in Nix. One can depend on all sorts of store objects with e.g. `builtins.storePath` --- getting around these distinction.

> including itself

A derivation never references itself. The outputs of a derivation however can reference themselves.

Reply:
So do you suggest removing the paragraph altogether?

Reply:
Heh I guess I didn't leave too much left. But maybe I can suggest and alternative.

Reply:
```suggestion
    Store objects can refer to both other store objects and themselves.
    References from a store object to itself are called *self-references*.
    References other than self-reference are not allowed to form a cycle.
```

Reply:
Maybe we can leave it out since that's essentially what will be in the architecture chapter on references anyway.